package cn.nitrowater.core.lib.infrastructure.security;

import io.jsonwebtoken.Claims;
import lombok.RequiredArgsConstructor;
import org.jetbrains.annotations.NotNull;
import org.springframework.core.convert.converter.Converter;
import org.springframework.security.oauth2.jwt.Jwt;
import org.springframework.security.oauth2.jwt.JwtDecoder;
import org.springframework.security.oauth2.jwt.JwtException;
import org.springframework.stereotype.Component;
import cn.nitrowater.core.lib.services.auth.AccessTokenService;

@Component
@Deprecated
@RequiredArgsConstructor
public class WaterJwtDecoder implements Converter<String, Jwt> , JwtDecoder {
    private final RsaJwtUtil rsaJwtUtil;
    private final AccessTokenService accessTokenServiceImpl;


    @Override
    public Jwt convert(@NotNull String token) {
            Claims claims = rsaJwtUtil.parseToken(token);
            accessTokenServiceImpl.validateAccessTokenAndRejectOld(claims);

            return Jwt.withTokenValue(token)
                    .header("alg","RS256")
                    .header("typ","JWT")
                    .claims(c -> c.putAll(claims))
                    .issuedAt(claims.getIssuedAt().toInstant())
                    .expiresAt(claims.getExpiration().toInstant())
                    .build();
    }

    @Override
    public Jwt decode(String token) throws JwtException {
        return this.convert(token);
    }
}
